[Nullability Annotations to Java Classes] Use Updated and Null Proof `MediaModel` Class (`breaking`)

ParaskP7 opened the pull request:
Parent [FluxC#2799](https://github.com/wordpress-mobile/WordPress-FluxC-Android/issues/2799)
Accompanying [FluxC#2886](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2886)

This PR accompanying the above FluxC PR, and uses an updated and null-proof [MediaModel.java](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/7fd7840a9bd900fde8e32ad8bd5632ad8f05dff3/fluxc/src/main/java/org/wordpress/android/fluxc/model/MediaModel.java#L4) class.

FYI: This change is `breaking`, meaning that any client that depended on the [MediaModel.java](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/7fd7840a9bd900fde8e32ad8bd5632ad8f05dff3/fluxc/src/main/java/org/wordpress/android/fluxc/model/MediaModel.java#L4) model should update to the new APIs (`non-default constructors`) as the existing API (`default constructor`) are now deprecated. As such, this change is inherently `risky`, meaning that there are compile-time changes associated with this change and thus needs testing to verify correctness, both on the library's and client's side.

-----

@AjeshRPai I added you as the main reviewer, not so randomly (it being a continuation of [FluxC#2878](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2878) and #19428), since I just wanted someone from the Jetpack/WordPress mobile team to be aware of and sign-off on that change. Feel free to merge this PR directly yourself if you deem so.

-----

FluxC Update List:

1. [Update fluxc version to pr hash (2886)](https://github.com/wordpress-mobile/WordPress-Android/commit/f6f93759b05bd0b1086cdeb7f3ec4042307ddcc4)
2. [Update fluxc version to pr hash (2886)](https://github.com/wordpress-mobile/WordPress-Android/pull/19506/commits/3b3e5a676e9c42bb0612220bda130e973c155770)
3. [Update fluxc version to trunk hash (2886)](https://github.com/wordpress-mobile/WordPress-Android/pull/19506/commits/5a9a89bffd684d8aca619589d52aa6d9ba460731)

Deprecation Resolution List (`breaking`):

1. [Resolve media model constructor deprecation warnings](https://github.com/wordpress-mobile/WordPress-Android/commit/1e8aa5a78b33b50211c970afe88df7433db4cfa0)
4. [Use one argument media error constructor on upload utils](https://github.com/wordpress-mobile/WordPress-Android/commit/a5f6bb43c57976a0734190f51cae271aa57dedc9)

Nullability Checks List:

1. [Un-guard usages of non-null media model url](https://github.com/wordpress-mobile/WordPress-Android/commit/fbcfd003a44c33e1560f3c59c8a07d5810ca01ab)
2. [Guard usages of nullable media model upload date](https://github.com/wordpress-mobile/WordPress-Android/commit/4d3b22ed0c3d7a882e49eca811cbcbe714779d99)
5. [Guard usages of nullable media model upload state](https://github.com/wordpress-mobile/WordPress-Android/commit/eff85644fd08806b6cfb561ab193fda79fefbb4a)
6. [Un-guard usages of non-null on media changed media list](https://github.com/wordpress-mobile/WordPress-Android/commit/d2ccda4114c8b9cb67ace356b75aa7602bc88239)
7. [Guard usages of nullable on media uploaded media](https://github.com/wordpress-mobile/WordPress-Android/commit/0c67de25930d539ae01bd519b52d7f220bc6c18d)
8. [Guard usages of nullable upload stock media error message](https://github.com/wordpress-mobile/WordPress-Android/commit/99c33d4c1b5ca2a9fac5843ecbff29b57f77229a)
9. [Guard usages of nullable get site media with id medial model](https://github.com/wordpress-mobile/WordPress-Android/commit/63112c92a121bcecb995b0d72428ad8eb78d863a)
10. [Guard usages of nullable get media with local id medial model](https://github.com/wordpress-mobile/WordPress-Android/commit/6a8cf682327b4ba8dca9f8d2229265ceff1e0e78)
11. [Un-guard usages of non-null get media for post with state list](https://github.com/wordpress-mobile/WordPress-Android/commit/db19fc36682dda988f2da463ede24dee3f8f5c53)

Warnings Suppression List:

1. [Suppress spread operator on page list event listener](https://github.com/wordpress-mobile/WordPress-Android/commit/bc63a428c06b4d21f9cd6f9d9ba53c9078efcd95)

-----

## To Test (`REST`):

- Smoke test any media related functionality on both, the WordPress and Jetpack apps, and see if everything is working as expected. For a couple of examples, you can expand and follow the inner and more explicit test steps within:

## To Test (`XMLRPC`):

- Create a new self-hosted WordPress site for XMLRPC testing purposes ([jurassic.ninja](https://jurassic.ninja/)).
- Smoke test any media related functionality on both, the WordPress and Jetpack apps, and see if everything is working as expected. For a couple of examples, you can expand and follow the inner and more explicit test steps within:

<details>
    <summary>Media Screens [MediaBrowserActivity.java + MediaGridFragment.kt + MediaSettingsActivity.java + MediaPreviewActivity.java + MediaPreviewFragment.java]</summary>

ℹ️ This test applies to both, the `Jetpack` and `WordPress` app.

- Go to `Media` screen, verify it is shown and functioning as expected.
- For example try changing tabs, searching and adding a new media.
- Tap on any media and verify that the `Media Settings` screen is shown and functioning as expected.
- For example try updating the media title or adding a media description.
- Tap on the media preview and verify that the `Media Preview` screen is shown and functioning as expected.
- For example try swipe left-right to navigate between media previews.

</details>

-----

## Merge instructions:

1. [x] Wait for the accompanying [FluxC#2886](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2886) PR to be merged.
2. [x] Update `wordPressFluxCVersion` to point to the `trunk` hash that includes the above solution.
3. [x] Remove the `[Status] Not Ready for Merge` label.
4. [x] Merge this PR.

-----

## Regression Notes

1. Potential unintended areas of impact

    - N/A

4. What I did to test those areas of impact (or what existing automated tests I relied on)

    - See `To test` section above.

5. What automated tests I added (or what prevented me from doing so)

    - N/A

-----

## PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

-----

## UI Changes testing checklist:

- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

ParaskP7 committed:
Build: Update fluxc version to pr hash (#2886)

This 'FluxC' PR hash updates the library to that branch version
where the 'MediaModel' is updated to its new null proof version.

FluxC PR: https://github.com/wordpress-mobile/
WordPress-FluxC-Android/pull/2886

This step is required in order to check that these 'FluxC'
related changes work as expected for WPAndroid.

ParaskP7 committed:
Analysis: Resolve media model constructor deprecation warnings

Warnings:
- Kotlin: "'constructor MediaModel()' is deprecated. Deprecated in Java"
- Java: "'MediaModel()' is deprecated"

ParaskP7 committed:
Analysis: Un-guard usages of non-null media model url

FYI: This change changes the logic to only checking for not blank
instead.

ParaskP7 committed:
Analysis: Guard usages of nullable media model upload date

ParaskP7 committed:
Analysis: Guard usages of nullable media model upload state

ParaskP7 committed:
Analysis: Un-guard usages of non-null on media changed media list

ParaskP7 committed:
Analysis: Suppress spread operator on page list event listener

Detekt Warning: "SpreadOperator: In most cases using a spread
operator causes a full copy of the array to be created before calling a
method. This may result in a performance penalty."

Now that this warning is suppressed close to source, it is no longer
needed to be in the baseline.

ParaskP7 committed:
Analysis: Guard usages of nullable on media uploaded media

ParaskP7 committed:
Analysis: Guard usages of nullable upload stock media error message

ParaskP7 committed:
Analysis: Use one argument media error constructor on upload utils

ParaskP7 committed:
Analysis: Guard usages of nullable get site media with id medial model

ParaskP7 committed:
Analysis: Guard usages of nullable get media with local id medial model

ParaskP7 committed:
Analysis: Un-guard usages of non-null get media for post with state list

ParaskP7 committed:
Merge branch 'trunk' of github.com:wordpress-mobile/WordPress-Android into analysis/use-updated-and-null-proof-media-model-class

# Conflicts:
#	build.gradle

ParaskP7 committed:
Build: Update fluxc version to pr hash (#2886)

This 'FluxC' PR hash updates the library to that branch version
where the 'MediaModel' is updated to its new null proof version.

FluxC PR: https://github.com/wordpress-mobile/
WordPress-FluxC-Android/pull/2886

This step is required in order to check that these 'FluxC'
related changes work as expected for WPAndroid.

------------------------------------------------------------------------

FYI: This change is an addition to
f6f93759b05bd0b1086cdeb7f3ec4042307ddcc4. This change is necessary in
order to overcome and resolve new merge conflicts.

ParaskP7 commented:
FYI: I'll resolve the `build.gradle` merge conflict when I'll be updating `wordPressFluxCVersion` to point to the `trunk` hash that includes the above solution (see `Merge instructions`).

AjeshRPai reviewed (approved):
Hey @ParaskP7 

I have tested the following scenarios and gone through the code changes. Everything looks good to me 👍🏼 
1. Publish a post and page with featured image + without featured image 
2. Publish a post and page with image and video 
3. Edit the image in a post and page
4. Upload image and video 
5. Add image from tenor library
6. Update the title, description and caption of an image and a video through media 
7. Download/Save the media to the device - from the detail screen of image/video 
8. Delete an image/video 

I am not sure if I have missed any other scenarios. I am approving but not merging so that you can take of care updating the fluxC version once the FluxC PR is merged.

ParaskP7 commented:
Awesome, thank for reviewing and testing this @AjeshRPai , you rock! 🙇 ❤️ 🚀 

FYI: I'll proceed with the merging myself. 🙏

ParaskP7 committed:
Merge branch 'trunk' of github.com:wordpress-mobile/WordPress-Android into analysis/use-updated-and-null-proof-media-model-class

# Conflicts:
#	build.gradle

ParaskP7 committed:
Build: Update fluxc version to trunk hash (#2886)

This 'FluxC' hash updates the library to that 'trunk' version
where the 'Media' model classes are updated to their new null proof
version.

FluxC PR: https://github.com/wordpress-mobile/
WordPress-FluxC-Android/pull/2886
